packages/react/dismissable-layer/src/DismissableLayer.tsx

```diff
@@ -1,241 +1,167 @@
+  (props, forwardedRef) => {
+    const {
+      disableOutsidePointerEvents = false,
+      onEscapeKeyDown,
+      onPointerDownOutside,
+      onFocusOutside,
+      onInteractOutside,
+      onDismiss,
+      ...layerProps
+    } = props;
+    const context = React.useContext(DismissableLayerContext);
+    const [node, setNode] = React.useState<DismissableLayerElement | null>(null);
+    const [, force] = React.useState({});
+    const composedRefs = useComposedRefs(forwardedRef, (node) => setNode(node));
+    const [highestModal] = [...context.modals].slice(-1);
+    const highestModalIndex = [...context.layers].indexOf(highestModal);
+    const layerIndex = node ? [...context.layers].indexOf(node) : -1;
+    const hasPointerEvents = layerIndex >= highestModalIndex;
+
+    const pointerDownOutside = usePointerDownOutside((event) => {
+      if (!hasPointerEvents) return;
       onPointerDownOutside?.(event);
       onInteractOutside?.(event);
-      if (!event.defaultPrevented) {
-        onDismiss?.();
-      }
-    }
-  });
+      if (!event.defaultPrevented) onDismiss?.();
+    });
 
-  // Dismiss on focus outside
-  const { onBlurCapture: handleBlurCapture, onFocusCapture: handleFocusCapture } = useFocusOutside(
-    (event) => {
+    const focusOutside = useFocusOutside((event) => {
       onFocusOutside?.(event);
       onInteractOutside?.(event);
-      if (!event.defaultPrevented) {
-        onDismiss?.();
-      }
-    }
-  );
+      if (!event.defaultPrevented) onDismiss?.();
+    });
 
-  // If we have disabled pointer events on body, we need to reset `pointerEvents: 'auto'`
-  // on some layers. This depends on which layers set `disableOutsidePointerEvents` to `true`.
-  //
-  // NOTE: it's important we set it on ALL layers that need it as we cannot simply
-  // set it on the deepest layer which sets `disableOutsidePointerEvents` to `true` and rely
-  // on inheritence. This is because layers may be rendered in different portals where
-  // inheritence wouldn't apply, so we need to set it explicity on its children too.
-  const isBodyPointerEventsDisabled = totalLayerCountWithDisabledOutsidePointerEvents > 0;
+    useEscapeKeydown((event) => {
+      const isHighestLayer = layerIndex === context.layers.size - 1;
+      if (!isHighestLayer) return;
+      onEscapeKeyDown?.(event);
+      if (!event.defaultPrevented) onDismiss?.();
+    });
+
+    useBodyPointerEvents({ disabled: disableOutsidePointerEvents });
+
+    React.useEffect(() => {
+      if (!node) return;
+      if (disableOutsidePointerEvents) context.modals.add(node);
+      context.layers.add(node);
+      dispatchUpdate();
+    }, [node, disableOutsidePointerEvents, context]);
+
+    React.useEffect(() => {
+      return () => {
+        if (!node) return;
+        context.layers.delete(node);
+        context.modals.delete(node);
+        dispatchUpdate();
+      };
+    }, [node, context]);
```

Comment:
There is no `DismissableLayer.Provider` anymore otherwise layers have to be part of the same React tree. Instead I mutate the `defaultContext` in `React.createContext` and dispatch an event on the document so that other dismissable layers know when to trigger a re-render.

This avoids us needing to expose a new `DismissableLayerProvider` part that consumers would need to wrap around their applications and I also think it could potentially lead to a solution to #1088 but will need to explore that separately.

Reply:
Curious why we need the context at all now. Could we track layer state locally and sync that via the event?

Reply:
ooooo interesting 👀

Reply:
looked into this and i'm not sure if it's worth it... i think the way it is currently makes it clearer that there is some state shared across primitives. thoughts?

![CleanShot 2022-02-15 at 12 06 17](https://user-images.githubusercontent.com/175330/154058970-eb2b49ca-c88f-44cc-98a1-ad10be1857b0.png)

Reply:
Yeah this feels a bit strange because then each instance of `DismissableLayer` is effectively tracking the state locally, duplicating it all over, and syncing it with the event. The context one feels more like a singleton which feels more appropriate here somehow.

Reply:
Yeah I see your point/s, losing that understanding for the sake of avoiding `force` probably isn’t worth it.